# riboseq - A Nextflow pipeline to perform Ribo-seq data analysis

## Table of contents

1. [Introduction](#introduction)
2. [Pipeline summary](#pipeline-summary)
3. [Quick start (testing)](#quick-start-testing)
4. [Quick start (running)](#quick-start-running)
5. [Pipeline parameters](#pipeline-parameters)
6. [Pipeline outputs](#pipeline-outputs)

## Introduction

riboseq is a Nextflow DSL2 pipeline for the analysis of Ribo-seq data.

## Pipeline summary

1. Adapter and quality trimming ([`Cutadapt`](https://cutadapt.readthedocs.io))
2. Premapping to remove small RNA mapping reads ([`bowtie2`]())
3. Mapping to genome and transcriptome ([`STAR`]())
4. UMI-based deduplication ([`UMI-tools`](https://umi-tools.readthedocs.io/en/latest/))

## Quick start (testing)

1. Ensure `Nextflow` and `Docker` or `Singularity` are installed on your system
2. Pull the main version of the pipeline from the GitHub repository:

```
nextflow pull ulelab/riboseq -r dev
```

3. Run the provided test dataset:

```
nextflow run ulelab/riboseq -r dev -profile conda,crick --org GRCh38
```

4. Review the results

## Quick start (running)

1. Ensure `Nextflow` and `Docker` or `Singularity` are installed on your system
2. Pull the main version of the pipeline from the GitHub repository:

```
nextflow pull ulelab/riboseq -r dev
```

3. Prepare a `samplesheet.csv` with your sample names and paths to your FASTQ files, following the template:

```
sample,fastq
sample1,/path/to/file1.fastq.gz
sample2,/path/to/file2.fastq.gz
sample3,/path/to/file3.fastq.gz
```

4. Run the pipeline (the minimum parameters have been specified):

```
nextflow run ulelab/riboseq -r dev \
-profile conda,crick \
--input samplesheet.csv \
--org GRCh38
```

## Pipeline parameters

### Profiles

- `-profile` can be used to specify `test`, `docker`, `singularity` and `crick` depending on the system being used and resources available. Others can be found at [nf-core](https://github.com/nf-core/configs).

### General parameters

- `--input` specifies the input sample sheet
- `--outdir` specifies the output results directory
    - default: `./results`
- `--tracedir` specifies the pipeline run trace directory
    - default: `./results/pipeline_info`

### Genome parameters

- `--org` specifies the organism (options are currently: `GRCh38`, `GRCm38`)


### Read trimming and alignment parameters



### Read mapping and UMI-based deduplication



### Optional pipeline modules

- `--skip_adapter_trimming` skips adapter and quality trimming
- `--skip_premap` skips premapping to the genome
- `--skip_qc` skips generation of QC plots and MultiQC report

## Pipeline outputs

The pipeline outputs results in a number of subfolders:

```
.
├── premap
├── map
├── deduplicated_genome
├── deduplicated_transcriptome
└── pipeline_info
```

### Files

- `premap` contains files resulting from alignment to the small RNA genome (smallrna_genome):
    - `*.bam` contains read alignments to the small RNA genome in BAM format
    - `*.bam.seqs.gz` contains the mapping locations and sequences of reads mapped to the small RNA genome
    - `*.unmapped.fastq.gz` contains the sequencing reads that did not map to the small RNA genome, in FASTQ format
    - `*.premap.log` is the bowtie2 log file
- `map` contains files resulting from alignment to the genome and transcriptome:
    - `*.Aligned.sortedByCoord.out.bam` contains read alignments to the genome in BAM format
    - `*.Aligned.toTranscriptome.out.bam` contains read alignments to the transcriptome in BAM format
    - `*.Log.final.out` is the STAR log output file
- `deduplicated_genome` contains files resulting after deduplication based on genomic location and UMIs:
    - `*.dedup.sorted.bam` contains the UMI deduplicated alignments to the genome in BAM format
    - `*.dedup.bed.gz` contains the the UMI deduplicated alignments to the genome in BED format
- `deduplicated_transcriptome` contains files resulting after deduplication based on transcriptomic location and UMIs:
    - `*.dedup.sorted.bam` contains the UMI deduplicated alignments to the transcriptome in BAM format
    - `*.dedup.bed.gz` contains the the UMI deduplicated alignments to the transcriptomie in BED format
- `pipeline_info` contains the execution reports, traces and timelines generated by Nextflow:
    - `execution_report.html`
    - `execution_timeline.html`
    - `execution_trace.txt`
